src/model/buffer.rs

```diff
@@ -72,267 +28,381 @@ pub struct Buffer {
+
+                self.num_newlines -= 1;
+                self.current_line -= 1;
+
+                let mut to_add: usize = 0;
+                for idx in (0..=self.node_list.index()).rev() {
+                    to_add += self.node_list.get(idx).offset()
+                        - self.node_list.get(idx).last_line_offset();
+                    if self.node_list.get(idx).has_newline() {
+                        break;
+                    }
+                }
+                self.cursor.line_offset = to_add;
+                self.cursor.original_line_offset = self.cursor.line_offset;
+            } else {
+                // deleted character should not be a newline
+                self.cursor.line_offset -= 1;
+                self.cursor.original_line_offset = self.cursor.line_offset;
             }
         } else {
             // cursor should be in the middle of a node
-            let current_node = &mut self.node_list[cursor.node_idx];
-            let from = current_node.from();
-            let index = current_node.index();
-            let offset = current_node.offset();
-            let line_offsets = current_node.line_offsets();
-
+            let from = self.node_list.get_curr().from();
+            let index = self.node_list.get_curr().index();
+            let offset = self.node_list.get_curr().offset();
+            let line_offsets = self.node_list.get_curr_mut().line_offsets();
+            let mut was_newline = false;
             // split the line offsets of the current node into two parts
+            let cursor_node_offset = self.cursor.node_offset;
             let mut left_line_offsets: Vec<usize> =
-                line_offsets.drain(..=cursor.line_idx).collect();
+                line_offsets.drain(..=self.cursor.line_idx).collect();
             let mut right_line_offsets: Vec<usize> = line_offsets
                 .drain(..)
-                .map(|x| x - cursor.node_offset)
+                .map(|x| x - cursor_node_offset)
                 .collect();
             right_line_offsets.insert(0, 0);
 
-            if *left_line_offsets.last().unwrap() == cursor.node_offset {
+            if self.cursor.line_offset == 0 {
                 // the last character of the left node should be '\n'
                 left_line_offsets.pop();
+                was_newline = true;
             }
 
-            let left_node = BufferNode::new(
-                from,
-                index,
-                index + cursor.node_offset - 1,
-                left_line_offsets,
-            );
+            let left_node =
+                BufferNode::new(from, index, self.cursor.node_offset - 1, left_line_offsets);
             let right_node = BufferNode::new(
                 from,
-                index + cursor.node_offset,
-                offset - cursor.node_offset,
+                index + self.cursor.node_offset,
+                offset - self.cursor.node_offset,
```

Comment:
Similar point here about `self.cursor.node_offset`, as it has been defined as `cursor_node_offset` in line 344